### PR TITLE
swarm: implemented retrieval test

### PR DESCRIPTION
**Pull request by holisticode**
Introduces retrieval tests on netstore

closes https://github.com/ethersphere/go-ethereum/issues/232

### CODE REVIEW

File: swarm/network/stream/common_test.go [hunk 1]
```diff
@@ -50,14 +50,20 @@ var (
...
-	registries       map[discover.NodeID]*TestRegistry
-	createStoreFunc  func(id discover.NodeID, addr *network.BzzAddr) (storage.ChunkStore, error)
+	defaultSkipCheck  bool
+	defaultDoRetrieve bool
+	waitPeerErrC      chan error
+	chunkSize         = 4096
+	registries        map[discover.NodeID]*TestRegistry
+	createStoreFunc   func(id discover.NodeID, addr *network.BzzAddr) (storage.ChunkStore, error)
+	getRetrieveFunc   = defaultRetrieveFunc
+	doRetrieve        = defaultDoRetrieve
```
Comment (zelig) [inline swarm/network/stream/common_test.go]: this is not used actually, so remove it and remove from defer or 
use it and skip the RPC subscribe call

File: swarm/network/stream/common_test.go [hunk 2]
```diff
@@ -90,19 +96,24 @@ func NewStreamerService(ctx *adapters.ServiceContext) (node.Service, error) {
...
 		waitPeerErrC <- waitForPeers(r, 1*time.Second, peerCount(id))
 	}()
-	dpa := storage.NewDPA(storage.NewNetStore(store, nil), storage.NewDPAParams())
+	dpa := storage.NewDPA(storage.NewNetStore(store, getRetrieveFunc(id)), storage.NewDPAParams())
 	testRegistry := &TestRegistry{Registry: r, dpa: dpa}
 	registries[id] = testRegistry
 	return testRegistry, nil
 }
 
+func defaultRetrieveFunc(id discover.NodeID) func(chunk *storage.Chunk) error {
```
Comment (zelig) [inline swarm/network/stream/common_test.go:111]: why is this needed?

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+				retrievalTest(t, c, n)
+			}
+		}
+	}
+}
+
+func retrievalTest(t *testing.T, chunkCount int, nodeCount int) {
+	//test live and NO history
+	log.Info("Testing live and no history", "chunkCount", chunkCount, "nodeCount", nodeCount)
+	live = true
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: why are these global and not params to `runRetrievalTest`?

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	conf.addrToIdMap = make(map[string]discover.NodeID)
+	conf.chunks = make([]storage.Key, 0)
+	//load nodes from the snapshot file
+	net, err := initNetWithSnapshot(nodeCount)
+	if err != nil {
+		return err
+	}
+	//do cleanup after test is terminated
+	defer func() {
+		doRetrieve = defaultDoRetrieve
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: this is not set to true so not needed

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+			if err != nil {
+				return err
+			}
+		}
+
+		//variables needed to wait for all subscriptions established before uploading
+		errc := make(chan error)
+
+		//now setup and start event watching in order to know when we can upload
+		ctx, watchCancel := context.WithTimeout(context.Background(), MAX_TIMEOUT*time.Second)
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: pls use go style constants

File: swarm/network/stream/common_test.go [hunk 6]
```diff
@@ -50,14 +50,20 @@ var (
...
 var (
-	defaultSkipCheck bool
-	waitPeerErrC     chan error
-	chunkSize        = 4096
-	registries       map[discover.NodeID]*TestRegistry
-	createStoreFunc  func(id discover.NodeID, addr *network.BzzAddr) (storage.ChunkStore, error)
+	defaultSkipCheck  bool
+	defaultDoRetrieve bool
+	waitPeerErrC      chan error
+	chunkSize         = 4096
```
Comment (zelig) [inline swarm/network/stream/common_test.go]: you reassign this later why global?
and anyway why are we using this?

File: swarm/network/stream/common_test.go [hunk 7]
```diff
@@ -50,14 +50,20 @@ var (
...
+	registries        map[discover.NodeID]*TestRegistry
+	createStoreFunc   func(id discover.NodeID, addr *network.BzzAddr) (storage.ChunkStore, error)
+	getRetrieveFunc   = defaultRetrieveFunc
```
Comment (zelig) [inline swarm/network/stream/common_test.go:63]: why need this?

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+kademlia network. The snapshot should have 'streamer' in its service list.
+*/
+func runRetrievalTest(chunkCount int, nodeCount int) error {
+	initRetrievalTest()
+	ids = make([]discover.NodeID, nodeCount)
+	disconnectC := make(chan error)
+	quitC := make(chan struct{})
+	conf = &synctestConfig{}
+	//map of discover ID to indexes of chunks expected at that ID
+	conf.idToChunksMap = make(map[discover.NodeID][]int)
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: where are all these used?

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+		//finally clear all data directories
+		datadirsCleanup()
+	}()
+	//get the nodes of the network
+	nodes := net.GetNodes()
+	//select one index at random...
+	idx := rand.Intn(len(nodes))
+	//...and get the the node at that index
+	//this is the node selected for upload
+	uploadNode := nodes[idx]
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go:523]: we should really modify this and upload to every node a file

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+			}
+			if healthy {
+				break
+			}
+		}
+
+		if history {
+			log.Info("Uploading for history")
+			//If testing only history, we upload the chunk(s) first
+			conf.chunks, err = uploadFileToSingleNodeStore(uploadNode.ID(), chunkCount)
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: we just need a hash and download the file by hash no? and really we should upload a file to each node and request it from every other node

File: swarm/network/stream/snapshot_retrieval_test.go [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+		default:
+		}
+		log.Trace(fmt.Sprintf("Checking node: %s", id))
+		//if there are more than one chunk, test only succeeds if all expected chunks are found
+		allSuccess := true
+
+		dpa := registries[id].dpa
+		for _, chnk := range conf.chunks {
+			reader := dpa.Retrieve(chnk)
+			if s, err := reader.Size(nil); err != nil || s != chunkSize {
```
Comment (zelig) [inline swarm/network/stream/snapshot_retrieval_test.go]: chunkSize is not needed and just use a hash of a file of potentially many chunks.
see janos PR

**Review by zelig — CHANGES_REQUESTED** [review]

File: swarm/network/stream/common_test.go (hunk 2 above)
Comment (holisticode) [reply swarm/network/stream/common_test.go:111]: Because the other tests prior to this retrieval test had the retrieval func set to `nil`, while we need to have the `retrieveFunc` set for the retrieval tests

File: swarm/network/stream/common_test.go (hunk 1 above)
Comment (holisticode) [reply swarm/network/stream/common_test.go]: removed

File: swarm/network/stream/snapshot_retrieval_test.go (hunk 3 above)
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go]: They need to be global due to the (current) `StartSyncing` test implementation, as they are needed there for setting up the stream, thus they were globally set in `snapshot_sync_test`. They could be additional RPC call params though.

File: swarm/network/stream/snapshot_retrieval_test.go (hunk 4 above)
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go]: removed

File: swarm/network/stream/common_test.go (hunk 7 above)
Comment (holisticode) [reply swarm/network/stream/common_test.go:63]: see above

File: swarm/network/stream/snapshot_retrieval_test.go (hunk 8 above)
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go]: Cleaned up. Some were remnants of earlier implementations.

File: swarm/network/stream/snapshot_retrieval_test.go (hunk 9 above)
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go:523]: This will be done in a new additional test

File: swarm/network/stream/snapshot_retrieval_test.go (hunk 10 above)
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go]: They are hashes. And yes, "upload file to each node and request it from every node" will be implemented as additional test.

File: swarm/network/stream/snapshot_retrieval_test.go (hunk 11 above)
Comment (holisticode) [reply swarm/network/stream/snapshot_retrieval_test.go]: this particular test only uploads single random chunks. will add additional test which will upload file(s) to different nodes

**Review by zelig — APPROVED** [review]

**Review by janos — APPROVED** [review]
